controller/tenant.go

```diff
@@ -2,314 +2,379 @@ package controller
+	// get user info
+	user, err := c.authClientService.GetUser(ctx)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{"err": err}, "creation of the user failed")
+		return jsonapi.JSONErrorResponse(ctx, err)
+	}
 
-	return openshift.RawUpdateTenant(ctx, openshiftConfig, t, envTypes, tenantService, allowSelfHealing)
-}
+	// gets tenant from DB
+	tenant, err := c.getExistingTenant(ctx, user.ID, user.OpenShiftUsername)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err":      err,
+			"tenantID": user.ID,
+		}, "retrieval of tenant entity from DB failed")
+		return jsonapi.JSONErrorResponse(ctx, errors.NewNotFoundError("tenants", user.ID.String()))
+	}
 
-// Clean runs the setup action for the tenant namespaces.
-func (c *TenantController) Clean(ctx *app.CleanTenantContext) error {
-	userToken := goajwt.ContextJWT(ctx)
-	if userToken == nil {
-		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError("Missing JWT token"))
+	// gets tenant's namespaces
+	namespaces, err := c.tenantRepository.GetNamespaces(user.ID)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err":      err,
+			"tenantID": user.ID,
+		}, "retrieval of existing namespaces from DB failed")
+		return jsonapi.JSONErrorResponse(ctx, err)
 	}
-	ttoken := &auth.TenantToken{Token: userToken}
 
-	// fetch the cluster the user belongs to
+	return ctx.OK(&app.TenantSingle{Data: convertTenant(ctx, tenant, namespaces, c.clusterService.GetCluster)})
+}
+
+// Update runs the update action.
+func (c *TenantController) Update(ctx *app.UpdateTenantContext) error {
+	// get user info
 	user, err := c.authClientService.GetUser(ctx)
 	if err != nil {
+		log.Error(ctx, map[string]interface{}{"err": err}, "creation of the user failed")
 		return jsonapi.JSONErrorResponse(ctx, err)
 	}
 
-	tenant, err := c.tenantService.GetTenant(ttoken.Subject())
+	// getting tenant from DB
+	dbTenant, err := c.getExistingTenant(ctx, user.ID, user.OpenShiftUsername)
 	if err != nil {
-		return jsonapi.JSONErrorResponse(ctx, errors.NewNotFoundError("tenants", ttoken.Subject().String()))
+		log.Error(ctx, map[string]interface{}{
+			"err":      err,
+			"tenantID": user.ID,
+		}, "retrieval of tenant entity from DB failed")
+		return jsonapi.JSONErrorResponse(ctx, errors.NewNotFoundError("tenant", user.ID.String()))
 	}
 
-	// restrict deprovision from cluster to internal users only
-	removeFromCluster := false
-	if user.UserData.FeatureLevel != nil && *user.UserData.FeatureLevel == "internal" {
-		removeFromCluster = ctx.Remove
-	}
+	TenantUpdater{Config: c.config, ClusterService: c.clusterService, TenantRepository: c.tenantRepository}.
+		Update(ctx, dbTenant, user, environment.DefaultEnvTypes, true)
 
-	cluster, err := c.clusterService.GetCluster(ctx, *user.UserData.Cluster)
+	ctx.ResponseData.Header().Set("Location", rest.AbsoluteURL(ctx.RequestData.Request, app.TenantHref()))
+	return ctx.Accepted()
+}
+
+type TenantUpdater struct {
```

Comment:
What do you by that? I don't follow... What would be the benefit?

Reply:
what `TenantService` you are referring to?

Reply:
oh, I thought there was already a `TenantService`. In anyways, this would probably be a good place for this code, I believe. But that could be done in a separate PR if you prefer